doc: power_optimization: expand docs (#12462)
nrfconnect/sdk-nrf: doc/nrf/test_and_optimize/optimizing/power_general.rst

.. note::
   Be careful with the use of pull-up resistors when designing the hardware for ultra-low power operation.
   An I/O pin with a 10 kΩ pull-up resistor that is set to ``GND`` will result in a current consumption of 300 µA at 3V.

Disable unused pins and peripherals
===================================

.. disable_unused_pins_start

Some of the pins and peripherals are enabled by default for some boards.
Depending on the peripheral or the pin type, they can increase the device power consumption to a different extent.
If the application does not use them, make sure they are disabled.

> To disable a particular peripheral, set its state in the board's :file:`dts` overlay to ``disabled``.

Reviewer: ```suggestion
To disable a particular peripheral, set its state in the board's devicetree overlay to ``disabled``.
```

For example, for **ADC**:

.. code-block:: devicetree

    &adc {
        status = "disabled";
    };

.. disable_unused_pins_end

Enable Device Power Management module
=====================================

.. enable_device_pm_start

The Device Power Management module provides an interface that the device drivers use to be informed about entering the suspend state or resuming from the suspend state.
This allows the device drivers to do any necessary power management operations, such as turning off device clocks and peripherals, which lowers the power consumption.

> To enable suspending peripherals when CPU goes to sleep, set the :kconfig:option:`CONFIG_PM_DEVICE` Kconfig option to ``y``.

Reviewer: ```suggestion
To enable suspending peripherals when the CPU goes to sleep, set the :kconfig:option:`CONFIG_PM_DEVICE` Kconfig option to ``y``.
```

.. enable_device_pm_end

Put the external flash into sleep mode in inactivity periods
============================================================

When the CPU goes to sleep, some of the peripherals are suspended by their drivers.

> However, the driver is not always able to know the application behavior and optimally handle the peripheral state.

Reviewer: ```suggestion
However, the driver is not always able to know the application behavior and handle the peripheral state optimally.
```
This is more a preference than a necessary change.

> One of such cases is the external flash usage by the applications.

Reviewer: ```suggestion
One such case is the external flash usage by the applications.
```

> It is typically used very rarely and only for the Device Firmware Upgrade purposes.

Reviewer: ```suggestion
It is very rarely used, and only for Device Firmware Upgrade purposes.
```
Alternatively:
```suggestion
It is typically not used, and only for Device Firmware Upgrade purposes.
```

Author: Applies v1.

> For this reason, you might want to suspend the external flash for majority of time and have it resumed to the active state only if needed.

Reviewer: ```suggestion
For this reason, you might want to suspend the external flash for the majority of the time and have it in active state only if needed.
```

> The Device Firmware Upgrade case is properly handled in the |NCS|, but for other proprietary use cases, you should handle state changes in your own implementation.

Reviewer: ```suggestion
The state change for the Device Firmware Upgrade case is handled in the |NCS|, but for other proprietary use cases, you should handle state changes in your own implementation.
```

> The Device Firmware Upgrade case is properly handled in the |NCS|, but for other proprietary use cases, you should handle state changes in your own implementation.

Reviewer 2: What do you mean by "properly handled?", that needs to be explained here, or provide a link which explains this.

Author: Thanks, @omkar3141 . I applied the suggestion from @wiba-nordic , and I think it solves the problem.

For example, to control the QSPI NOR external flash, you can use the following implementation:

.. code-block:: C++

    #include <zephyr/pm/device.h>

    const auto * qspi_dev = DEVICE_DT_GET(DT_INST(0, nordic_qspi_nor));
    if (device_is_ready(qspi_dev))
    {
        // Put the peripheral into suspended state.
        pm_device_action_run(qspi_dev, PM_DEVICE_ACTION_SUSPEND);

        // Resume the peripheral from the suspended state.
        pm_device_action_run(qspi_dev, PM_DEVICE_ACTION_RESUME);
    }

> 

Reviewer: ```suggestion

Power down unused RAM
=====================

:ref:`Optimizing the memory footprint <app_memory>` can reduce the amount of RAM used, and powering down the unused RAM reduces power consumption.
In the |NCS|, you can use the :ref:`lib_ram_pwrdn` library to power down the unused RAM.

```
I am not 100% happy with this suggestion.

Author: How can I make you 100% happy?

Reviewer: ```suggestion

Power down unused RAM
=====================

:ref:`Optimizing the memory footprint <app_memory>` can reduce the amount of RAM used, but unused RAM still consumes power.
You can power down this unused RAM to reduce the power consumption of your application.
In the |NCS|, you can use the :ref:`lib_ram_pwrdn` library to power down unused RAM.

```
That's a lot of RAM, but I'm much happier with this.

Author: Rephrased a bit.

Configure radio transmitter power
=================================

.. radio_power_start

The radio transmitter power (radio TX power) has a significant impact on the device power consumption.

> The higher the transmitting power, the greater the wireless communication range, which leads to higher power consumption.

Reviewer: ```suggestion
The higher the transmitter power, the greater the wireless communication range, but it also leads to higher power consumption.
```
This is the only line with "transmitting". I'd recommend using the same word for the whole section, even though both words are grammatically correct.

Make sure to choose the optimal configuration for your specific use case.

.. radio_power_end

.. _app_power_opt_protocols:

Protocol-specific recommendations
*********************************

Besides applying `General recommendations`_, read the following subsections for more information on how to optimize specific subsystems.

Bluetooth mesh
==============

The Bluetooth mesh protocol offers the :ref:`ug_bt_mesh_configuring_lpn` feature for optimizing the power consumption of the Bluetooth mesh devices.

Matter
======

To optimize the power consumption of your Matter application, complete the actions listed on the :ref:`ug_matter_device_low_power_configuration` page.

Thread
======

> The Thread protocol offers the :ref:`Sleepy End Device types <thread_sed_ssed>` feature for optimizing the power consumption of the Thread devices.

Reviewer: @edmont please verify that this is correct information, and if there is anything to add.

> The Thread protocol offers the :ref:`Sleepy End Device types <thread_sed_ssed>` feature for optimizing the power consumption of the Thread devices.

Reviewer 3: I would rater state something like:

> Take a look to :ref:`ot_cli_sample_low_power` on how the :ref:`ot_cli_sample` sample can be configured following some of the general recommendations to achieve a low power consumption operation and perform power consumption measurements.

> The Thread protocol offers the :ref:`Sleepy End Device types <thread_sed_ssed>` feature for optimizing the power consumption of the Thread devices.

Reviewer: Amending my suggestion based on that suggestion:
```suggestion
The Thread protocol offers :ref:`Sleepy End Device types <thread_sed_ssed>` for optimizing the power consumption of the Thread devices.
Sleepy End Devices try to limit their power consumption by sleeping most of the time.

The :ref:`ot_cli_sample` sample can be used to perform power consumption measurements when configured following some of the general recommendations.
See the :ref:`ot_cli_sample_low_power` section of the sample documentation for more information.
```

Wi-Fi
=====

> The Wi-Fi protocol introduces the power save mechanism that allows the Station device (STA) to spend the majority of time in a sleep state and wake-up periodically to check for pending traffic.

Reviewer: ```suggestion
The Wi-Fi protocol introduces the power save mechanism that allows the Station device (STA) to spend the majority of the time in a sleep state and wake-up periodically to check for pending traffic.
```

> The Wi-Fi protocol introduces the power save mechanism that allows the Station device (STA) to spend the majority of time in a sleep state and wake-up periodically to check for pending traffic.

Reviewer 4: ```suggestion
The Wi-Fi protocol introduces the power save mechanism that allows the Station (STA) device to spend the majority of time in a sleep state and wake up periodically to check for pending traffic.
```

Reviewer 4: > 

@sachinthegreen @bama-nordic : Can you please take a look at the Wi-Fi section here to verify the info and see if anything else is required here?

For more information about the Wi-Fi power save mechanism, see the :ref:`Wi-Fi MAC layer <wifi_mac_layer>` documentation.

> To enable the Wi-Fi power save mode, set the :kconfig:option:`CONFIG_NRF_WIFI_LOW_POWER` Kconfig option to ``y``.

Reviewer 4: Please add this after line 193:

See :ref:`ug_nrf70_developing_powersave` for more information about the power save modes.
